lib/sendgrid_actionmailer.rb

```diff
@@ -9,13 +9,14 @@ class DeliveryMethod
 
     include SendGrid
 
-    attr_reader :client, :raise_delivery_errors
+    DEFAULTS = {
+      raise_delivery_errors: false,
+    }
 
-    def initialize(params)
-      # SendGrid::API is a wrapper of that...
-      # https://github.com/sendgrid/ruby-http-client/blob/master/lib/ruby_http_client.rb
-      @client = SendGrid::API.new(api_key: params.fetch(:api_key)).client
```

Comment:
@h6ah4i Why remove @client as an attr_accessor set on initialization?  With this change, client is now only accessible within perform_send_request and SendGrid::API.new(...).client must be invoked upon every perform_send_request execution?

Reply:
Thanks for your advice. I moved the client initialization code here, because the [official SMTP implementation](https://github.com/mikel/mail/blob/master/lib/mail/network/delivery_methods/smtp.rb) seems that it considers `settings` attirbute can be modified anytime.

However, that could be my over-thinking. I checked some thirdparty ActionMailer integrations and found that they are caching their API client like this;

```ruby
  def client
    @client ||= ApiClient.new
  end
```

- https://github.com/jorgemanrubia/mailgun_rails/blob/master/lib/mailgun_rails/deliverer.rb
- https://github.com/zaeznet/elastic_email_rails/blob/master/lib/elastic_email/deliverer.rb

Which is better to revert the move or caching?

---

P.S. Do you know any official guidance or tutorial how to create a custom deliverer? Thanks.

Reply:
I think either works and both are preferable to having client initialization buried within the sending method. The following is a pretty standard ruby paradigm:

```
  def client(**params)
    @client ||= ApiClient.new(params)
  end
```

If you want to make that change and resolve the spec conflict, I'll merge this in.

I do question the benefit of changing settings and modifying client within a live app.  The only benefit I can see is perhaps the ability to send across multiple api keys? But if that were the goal, I think there are probably better ways than modifying the settings on an initialized object in a production app.   ActionMailer does provide a way of [Sending Emails with Dynamic Delivery Options](https://guides.rubyonrails.org/action_mailer_basics.html#sending-emails-with-dynamic-delivery-options). Perhaps the gem could utilize this same approach, by passing the api_key as part of the delivery_method_options to mail?

Reply:
> If you want to make that change and resolve the spec conflict, I'll merge this in.

Thanks, I'll fix them soon.

> I do question the benefit  ...  as part of the delivery_method_options to mail?

That makes sense. ActionMailer creates a new DeliveryMethod instance on every `mail()` method calls. So we do not have to take care about reusing DeliveryMethod anymore in live apps.